crates/bevy_asset/src/io/wasm_asset_io.rs

```diff
@@ -0,0 +1,54 @@
+use crate::{AssetIo, AssetIoError};
+use anyhow::Result;
+use async_trait::async_trait;
+use js_sys::Uint8Array;
+use std::path::{Path, PathBuf};
+use wasm_bindgen::JsCast;
+use wasm_bindgen_futures::JsFuture;
+use web_sys::Response;
+
+pub struct WasmAssetIo {
+    root_path: PathBuf,
+}
+
+impl WasmAssetIo {
+    pub fn new<P: AsRef<Path>>(path: P) -> Self {
+        WasmAssetIo {
+            root_path: path.as_ref().to_owned(),
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl AssetIo for WasmAssetIo {
+    async fn load_path(&self, path: &Path) -> Result<Vec<u8>, AssetIoError> {
+        let path = self.root_path.join(path);
+        let window = web_sys::window().unwrap();
```

Comment:
This will not work in web workers and `WorkerGlobalScope` should be used instead. However, `web-sys` doesn't seem to provide any abstraction to access this cross-environments as `WindowOrWorkerGlobalScope` mixin is not supported.

I think the best option would be to access `js_sys::global()` and try casting it to either `Window` or `WorkerGlobalScope`, whichever succeeds. Similar fix was done in https://github.com/rustwasm/gloo/pull/106, but it's kind of ugly. Maybe it could land in `bevy_utils`?

Reply:
haha anything specific about web-sys / webworkers is going to be out of my depth at the moment. but that sounds reasonable to me. given that we know this works outside of webworkers, im inclined to merge as-is for now. Then someone who knows what they're doing (like you) can follow up with webworker compat.

Reply:
> Regarding `block_on`, I think it would be best to make loaders async as it is the only way to make it work on single threaded wasm. GLTF loading with sub assets would still be broken, but that can be fixed by making schedule async as in the wasm threads PR.

+1 for making loaders async - I did it on my branch and it was very simple change

Reply:
I was hesitant to make asset loaders async because I didn't want consumers to need the `#[async_trait]` proc-macro. But:

1. atelier-assets uses async loaders
2. they do it without requiring `#[async_trait]` in consumers by using BoxFuture
3. even if async_trait was required, thats not a huge loss relative to what we get

I'll make the change

Reply:
here is my implementation, for example working gltf loader: https://github.com/mrk-its/bevy/commit/05bf8aeb3a48abdf3e17c014dbabaed28833293f